### ISSUE TITLE
sh2power is wrong (I think)

### ISSUE THREAD

**Issue by dchristiaens** [MRtrix3/mrtrix3#530]
Say I compute the RMS amplitude of some SH function as
```
sh2amp sh.mif dirs500.txt - | mrcalc - - -mult - | mrmath - mean - -axis 3 | mrcalc - -sqrt rms0.mif
```
Normally, one should be able to compute the same thing with `sh2power`, i.e.,
```
sh2power sh.mif - | mrmath - sum - -axis 3 | mrcalc - -sqrt rms1.mif
```
However, I noticed that these have different contrasts:
![rms0.mif](https://cloud.githubusercontent.com/assets/7466983/14423407/0022aef2-ffdc-11e5-8927-5fde0fedd627.png)
![rms1.mif](https://cloud.githubusercontent.com/assets/7466983/14423435/1b246c9a-ffdc-11e5-9cc7-31d18dcdfe78.png)

However, if I compute the root-sum-squared of all SH coefficients, divided by sqrt(4pi) as such 
```
mrcalc sh.mif sh.mif -mult - | mrmath - sum - -axis 3 | mrcalc - 12.5664 -divide -sqrt rms2.mif
```
then I do get the correct result:
![rms2.mif](https://cloud.githubusercontent.com/assets/7466983/14423494/666463f4-ffdc-11e5-9fcf-4f0ae740001b.png)

The difference is of course the division with (2*l+1) in sh2power on [line 63](https://github.com/MRtrix3/mrtrix3/blob/master/cmd/sh2power.cpp#L63). Therefore, I think this division should be deleted, unless I'm missing something?

What I think happened here is that this originates from the legacy SH basis implementation, which used different normalization. The new orthonormal basis is then supposedly normalized per SH order l, rather than for each individual (l,m) basis function.

I take at least part of the blame on this one, since I'm the one who ported `sh2power` to the new syntax and wrote the test for it. At the time, I only tested whether its output was consistent between the old and new syntax...

**Comment by jdtournier** [comment]
Well, it's both right and wrong, depends on how you view it... The correction of the SH basis is done for the m!=0 terms, and you can see it being applied a few lines earlier. The idea behind the division by (2*l+1) is to produce the power _per coefficient_, rather than as a total. I can't remember what the rationale was for that decision... But it would mean that the sum of the powers wouldn't match the expected value for the total power. 

I'm not sure there is any reason not to remove that normalisation step - it is indeed counter-intuitive when presented that way. Can anyone think of an uses for `sh2power` currently, and whether these would be affected by making the suggested change?

**Comment by jdtournier** [comment]
By the way, I am surprised to find that this:

    sh2amp sh.mif dirs500.txt - | mrcalc - - -mult - | ...

worked at all: the backend is supposed to read the next line of input as the filename for the next piped argument... In your case, it would have been cleaner to write:

    sh2amp sh.mif dirs500.txt - | mrcalc - 2 -pow - | ...

But I'm surprised your original command worked at all, given [how it's handled in the code](https://github.com/MRtrix3/mrtrix3/blob/master/lib/formats/pipe.cpp#L29-L33). Seems `getline()` returns the same value twice?? Why wouldn't this lead to a double deletion of the temporary file at exit? 

Actually, it more likely to be due to the [feature I added to `mrcalc` to prevent double-loading of the same image when used repeatedly in the expression](https://github.com/MRtrix3/mrtrix3/blob/master/cmd/mrcalc.cpp#L223-L228)... Not sure that's the expected behaviour, but I can't see it causing issues very often - it would only matter for people piping multiple images from a single command to the next command, which I've never seen anyone attempt (myself included)...

**Comment by dchristiaens** [comment]
> I can't remember what the rationale was for that decision...

It might have to do with [Wikipedia](https://en.wikipedia.org/wiki/Spherical_harmonics#Spectrum_analysis), which describes Parseval's theorem for Schmidt semi-normalized harmonics. That formula is not valid for the Descoteaux basis, however, the difference being the (2l+1) normalization I mentioned...

In general, I also find the syntax of `sh2power` a bit confusing. Could we at least include an option to output a 3D image of the full power, as this is presumably what people will need most of the time? Personally, I would even make that the default and optionally output the power-per-order.

> By the way, I am surprised to find that this:
> ```
> sh2amp sh.mif dirs500.txt - | mrcalc - - -mult - | ...
> ```
> worked at all

Well, it's not a typo... I use this regularly and it does work fine on my system. Somehow I have completely overlooked the `-pow` operator so far. Thanks for pointing me to it -- it is indeed more clean with piped IO.

**Comment by jdtournier** [comment]
> Could we at least include an option to output a 3D image of the full power, as this is presumably what people will need most of the time? Personally, I would even make that the default and optionally output the power-per-order.

You're probably right, that would indeed make the most sense... I personally tend to look at the power per harmonic band (I find the power 2 band makes a nice substitute of the FA...), but I do agree that the default behaviour should probably be to do what the command name implies...

**Comment by thijsdhollander** [comment]
I typically use it by default in it's original form as well (quite often), to investigate the prevalence of the (need for) certain orders spatially.  The name is of course an abbreviation for something like "power spectrum"; but the documentation is indeed severely lacking.  I think it was originally implemented (by @Lestropie ?) for the purpose of Flavio's T1-like map presented at ISMRM a couple of years ago... but I might be mistaken here.  Anyone who remembers that one?

If people need the RMS, they can indeed quite easily obtain that using `mrmath` and `mrcalc`... no?  After all its just squaring, averaging, and taking the root; as the name implies. :smile: 

Anyways, not against including an option in `sh2power` for it as well; but I wouldn't really make it the default... the version to get the power spectrum / power per order is the one with the most added value (much more hassle to obtain via piping and combining outputs from other commands).  If anything, the documentation should be made much more clear on what is actually being computed.

**Comment by Lestropie** [comment]
>  I think it was originally implemented (by @Lestropie ?) for the purpose of Flavio's T1-like map presented at ISMRM a couple of years ago... but I might be mistaken here. Anyone who remembers that one?

No, the `sh2power` command's been around for a long time, well before that: `git blame` reports late 2010. I was going to include Flavio's APM as part of `sh2metric` (#66) but that got canned.

**Comment by thijsdhollander** [comment]
Aha, good to know.  Yeah, that's going back quite a bit further then.  Found the power maps thing again btw: http://submissions.mirasmart.com/ismrm2014/proceedings/files/0730.pdf

**Comment by Lestropie** [comment]
> Found the power maps thing again btw:

I still have code for generating those maps in my personal repo. Just never made any use of it.

**Comment by jdtournier** [comment]
Yes, `sh2power` is one of mine... Personally I agree with @dchristiaens that given the command name is `sh2power`, its output should by default be the power in the series. I don't think there's anything in any of our scripts or tutorials that currently relies on `sh2power` (although that might change at some point). So I'm more than happy to default to the full power as a 3D image, with no normalisation by (2*l+1). I like the notion of a _spectrum_ though (thanks @thijsdhollander), so I vote the option is call `-spectrum` to provide the per-harmonic order power spectrum. I also don't think we should be normalising by (2*l+1) in this case either, technically that wouldn't actually correspond to the power in that band... And I can't think of any applications where that might make a difference currently... I agree with @thijsdhollander that for me, the most common use case will be to get the specturm, but it's not that much hassle to add `-sp` to the command... 

Anyway, I'm happy either way, as long as it's better documented than when I wrote it... and the normalisation by (2*l+!) is gone, I really can't think of a use-case where it makes sense.

**Comment by dchristiaens** [comment]
Darn, I was half-way through a post saying exactly the same until @jdtournier beat me to it. And it's one hour earlier for him... :-)

So in summary, I don't care much about the default but I do think that the (2L+1)-normalisation needs to go. My initial option name `-full` was indeed a terrible choice, and I also think @thijsdhollander's idea of `-spectrum` is great.

**Comment by thijsdhollander** [comment]
Ok, I can live with that. ;-) :+1: 

...but the documentation on this one is crucial indeed.

**Comment by jdtournier** [comment]
Closed with #532.
